### ISSUE TITLE
Azure App Service step fails to deploy FunctionApp when using a Linux App Service Plan

### ISSUE THREAD

**Issue by mjhilton** [OctopusDeploy/Issues#7357]
### Team

- [X] I've assigned a team label to this issue

### Severity

Discovered internally, no customer reports as yet

### Version

Current

### Latest Version

I could reproduce the problem in the latest build

### What happened?

Deploying a FunctionApp application using the Azure App Service step fails when deploying to a Linux FunctionApp with a Linux AppServicePlan.

### Reproduction

1. In the Azure Portal, create a Linux Function App on .NET 6, with an App Service Plan of tier B1, also running Linux
2. Create a Project in Octopus which uses the `Azure App Service` step, to attempt to deploy a .NET FunctionApp package to the Function App

**Expected:** Octopus should be able to deploy to the function app as it does when the App Service Plan is Windows
**Actual**: Deployment fails.

### Error and Stacktrace

```shell
5:53:03   Verbose  |       Using Calamari.AzureAppService.win-x64 2.1.1
05:53:04   Verbose  |       Acquiring isolation mutex RunningScript with NoIsolation in ServerTasks-15893
05:53:04   Verbose  |       Executable directory is C:\Windows\system32\WindowsPowershell\v1.0
05:53:04   Verbose  |       Executable name or full path: C:\Windows\system32\WindowsPowershell\v1.0\PowerShell.exe
05:53:04   Verbose  |       No user context provided. Running as current user.
05:53:06   Verbose  |       Starting C:\Windows\system32\WindowsPowershell\v1.0\PowerShell.exe in working directory 'C:\Octopus\Tentacle\Work\20220204055303-15893-16' using 'OEM United States' encoding running as 'NT AUTHORITY\SYSTEM' with the same environment variables as the launching process
05:53:06   Verbose  |       Calamari Version: 1.0.0
05:53:06   Verbose  |       Environment Information:
05:53:06   Verbose  |       OperatingSystem: Microsoft Windows NT 10.0.17763.0
05:53:06   Verbose  |       OsBitVersion: x64
05:53:06   Verbose  |       Is64BitProcess: True
05:53:06   Verbose  |       CurrentUser: NT AUTHORITY\SYSTEM
05:53:06   Verbose  |       MachineName: WIN-QH0AA0NEJ1D
05:53:06   Verbose  |       ProcessorCount: 2
05:53:06   Verbose  |       CurrentDirectory: C:\Octopus\Tentacle\Work\20220204055303-15893-16
05:53:06   Verbose  |       TempDirectory: C:\Windows\TEMP\
05:53:08   Verbose  |       HostProcess: Calamari.AzureAppService (4048)
05:53:10   Verbose  |       Extracting package to: C:\Octopus\Tentacle\Work\20220204055303-15893-16\staging
05:53:10   Verbose  |       Extracted 9 files
05:53:23   Verbose  |       Deployment type: Package
05:53:23   Info     |       Uploading package to abc-linux-functionapp
05:53:23   Verbose  |       Path to upload: C:\Octopus\Tentacle\Files\linux-functionapp@S1.0.0@41EFF29A1F5DDC4EB1111893F690D888.zip
05:53:23   Verbose  |       Target Site: abc-linux-functionapp
05:53:55   Verbose  |       Publishing C:\Octopus\Tentacle\Files\linux-functionapp@S1.0.0@41EFF29A1F5DDC4EB1111893F690D888.zip to https://mjh-linux-functionapp.scm.azurewebsites.net:443/api/zipdeploy
05:53:55   Error    |       Running rollback behaviours...
05:53:55   Error    |       Bad Request
05:53:55   Error    |       System.Exception
05:53:55   Error    |       at Calamari.AzureAppService.Behaviors.AzureAppServiceBehaviour.UploadZipAsync(PublishingProfile publishingProfile, HttpClient client, String uploadZipPath, String targetSite) in C:\buildAgent\work\cdb95c8a359b9bc9\source\Calamari\Behaviors\AzureAppServiceBehaviour.cs:line 155
05:53:55   Error    |       at Calamari.AzureAppService.Behaviors.AzureAppServiceBehaviour.Execute(RunningDeployment context) in C:\buildAgent\work\cdb95c8a359b9bc9\source\Calamari\Behaviors\AzureAppServiceBehaviour.cs:line 114
05:53:55   Error    |       at Calamari.Common.Plumbing.Pipeline.PipelineCommand.ExecuteBehaviour(RunningDeployment context, IBehaviour behaviour)
05:53:55   Error    |       at Calamari.Common.Plumbing.Pipeline.PipelineCommand.Execute(ILifetimeScope lifetimeScope, IVariables variables)
05:53:55   Error    |       at Calamari.Common.Plumbing.Pipeline.PipelineCommand.Execute(ILifetimeScope lifetimeScope, IVariables variables)
05:53:55   Error    |       at Calamari.Common.CalamariFlavourProgramAsync.Run(String[] args)
05:53:55   Verbose  |       Process C:\Windows\system32\WindowsPowershell\v1.0\PowerShell.exe in C:\Octopus\Tentacle\Work\20220204055303-15893-16 exited with code 100
05:53:55   Verbose  |       Updating manifest with output variables
05:53:55   Verbose  |       Updating manifest with action evaluated variables
05:53:55   Fatal    |       The remote script failed with exit code 100
05:53:55   Fatal    |       The action Deploy FunctionApp on Linux Functionapp failed
```

### More Information

This is the same root cause as other issues that have been raised against the corresponding [AzureDevOps Pipelines Task](https://github.com/microsoft/azure-pipelines-tasks/issues/8657#issuecomment-432530739) and the [JetBrains Azure Tools](https://github.com/JetBrains/azure-tools-for-intellij/issues/271)

The core of the problem is that when running a Linux AppServicePlan, ZipDeploy requires the `Content-Type` header in the file upload request to have a value of `application/octet-stream`. When hosting on a Windows AppServicePlan, the Windows version of the deployment endpoint seems to be more forgiving and accepts Zip uploads without a `Content-Type` header.

I've been able to replicate the problem in an automated test, and found that updating the `AzureAppServiceBehaviour` to explicitly [set the ContentType when posting the ZIP file to Kudu](https://github.com/OctopusDeploy/Sashimi.AzureAppService/blob/main/source/Calamari/Behaviors/AzureAppServiceBehaviour.cs#L152-L153) will fix the problem.

![image](https://user-images.githubusercontent.com/730704/152490829-d513e209-c58b-411c-b4c5-ef06284594a5.png)

### Workaround

Host your FunctionApp on a Windows-based AppServicePlan instead

**Comment by mjhilton** [comment]
And the start of a new test for `Calamari.Tests.AppServiceBehaviourFixture` that I wrote when trying to replicate this problem. Needs a couple more tweaks in the ASP and Site properties to get it behaving correctly, but most of it is there:

```
        [Test]
        public async Task CanDeployZip_ToLinuxFunctionApp()
        {
            var (_, linuxFunctionAppName) = await CreateLinuxFunctionApp();
            var packageinfo = PrepareZipPackage();

            var result = await CommandTestBuilder.CreateAsync<DeployAzureAppServiceCommand, Program>().WithArrange(context =>
            {
                context.WithPackage(packageinfo.packagePath, packageinfo.packageName, packageinfo.packageVersion);
                context.Variables.Add(AccountVariables.ClientId, clientId);
                context.Variables.Add(AccountVariables.Password, clientSecret);
                context.Variables.Add(AccountVariables.TenantId, tenantId);
                context.Variables.Add(AccountVariables.SubscriptionId, subscriptionId);
                context.Variables.Add("Octopus.Action.Azure.ResourceGroupName", resourceGroupName);
                context.Variables.Add("Octopus.Action.Azure.WebAppName", linuxFunctionAppName);
                context.Variables.Add(SpecialVariables.Action.Azure.DeploymentType, "ZipDeploy");
                context.AddVariable(AccountVariables.Environment, "AzureCloud");
            }).Execute();
            
            await AssertContent($"{linuxFunctionAppName}.azurewebsites.net", $"Hello {greeting}");
        }

        private async Task<(string appServicePlan, string functionApp)> CreateLinuxFunctionApp()
        {
            var linuxSvcPlan = await webMgmtClient.AppServicePlans.BeginCreateOrUpdateAsync(resourceGroupName,
                $"{resourceGroupName}-linux-asp",
                new AppServicePlan(resourceGroupLocation) {Sku = new SkuDescription("B1", "Basic"), Kind = "linux", Reserved = true }
            );
            
            var functionAppSite = await webMgmtClient.WebApps.BeginCreateOrUpdateAsync(resourceGroupName,
                $"{resourceGroupName}-linux",
                new Site(resourceGroupLocation)
                {
                    ServerFarmId = linuxSvcPlan.Id, Kind = "functionapp,linux", Reserved = true, SiteConfig = new SiteConfig
                    {
                        AlwaysOn = true,
                        LinuxFxVersion = "DOTNET|6.0",
                        Use32BitWorkerProcess = true
                    }
                }
            );
```

**Comment by mjhilton** [comment]
Adding an [🔒 internal report](https://octopusdeploy.slack.com/archives/C020P7U9Z9S/p1648103265860369) and +1 to getting this done

**Comment by nathanwoctopusdeploy** [comment]
Experiencing a similar deployment error with a Linux App Service when the app is set to Run from Zip https://github.com/OctopusDeploy/Issues/issues/7456

**Comment by mjhilton** [comment]
I'm no longer able to reproduce this error using the reproduction steps as logged in main issue body.

As @nathanwoctopusdeploy mentions, there's a reliable reproduction of the same symptoms when a Linux-based the FunctionApp has the `WEBSITE_RUN_FROM_PACKAGE` flag turned on. 

I believe that the initial reproduction I did may have had that flag turned on (or otherwise something has changed in The Matrix and the Kudu team has made the non-run-from-package version more flexible).

Closing this one as unable to reproduce, and will push my fix against #7456 with tests for both values of the flag to prevent regression.
